File: Leetcode Daily Challenge/025-July-2025/018-minimum-difference-in-sums-after-removal-of-elements.java
// https://leetcode.com/problems/minimum-difference-in-sums-after-removal-of-elements/




import java.util.*;

class Solution {
    // TC : O(N*log(N))
    // SC : O(N)
    public long minimumDifference(int[] nums) {
        int n = nums.length;
        int len = n / 3;

        long[] leftMin  = new long[n];
        long[] rightMax = new long[n];
        PriorityQueue<Integer> minH = new PriorityQueue<>();
        PriorityQueue<Integer> maxH = new PriorityQueue<>(Collections.reverseOrder());
        
        long leftMinSum = 0, rightMaxSum = 0;
        for (int i = 0; i < 2*len; i ++) {
            leftMinSum += nums[i];
            maxH.offer(nums[i]);
            if (maxH.size() > len) {
                leftMinSum -= maxH.poll();
            }
            leftMin[i] = leftMinSum;

            rightMaxSum += nums[n - i - 1];
            minH.offer(nums[n - i - 1]);
            if (minH.size() > len) {
                rightMaxSum -= minH.poll();
            }
            rightMax[n - i - 1] = rightMaxSum;
        }

        long result = Long.MAX_VALUE;
        for (int i = len - 1; i < 2*len; i ++) {
            result = Math.min(result, (leftMin[i] - rightMax[i + 1]));
        }
        return result;
    }
}

File: Leetcode Daily Challenge/025-July-2025/018-minimum-difference-in-sums-after-removal-of-elements.cpp
// https://leetcode.com/problems/minimum-difference-in-sums-after-removal-of-elements/




#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    using ll = long long;

    // TC : O(N*log(N))
    // SC : O(N)
    long long minimumDifference(vector<int>& nums) {
        int n = nums.size();
        int len = n/3;

        vector<ll> leftMin(n, 0), rightMax(n, 0);
        priority_queue<ll> maxH;
        priority_queue<ll, vector<ll>, greater<>> minH;

        ll sumLeftMin = 0, sumRightMax = 0;
        for (int i = 0; i < 2*len; i ++) {
            sumLeftMin += nums[i];
            maxH.push(nums[i]);
            if (maxH.size() > len) {
                sumLeftMin -= maxH.top();
                maxH.pop();
            }
            leftMin[i] = sumLeftMin;

            sumRightMax += nums[n - i - 1];
            minH.push(nums[n - i - 1]);
            if (minH.size() > len) {
                sumRightMax -= minH.top();
                minH.pop();
            }
            rightMax[n - i - 1] = sumRightMax;
        }

        ll result = LONG_MAX;
        for (int i = len - 1; i < 2*len; i ++) {
            result = min(result, (leftMin[i] - rightMax[i + 1]));
        }
        return result;
    }
};